composer/trainer/checkpoint.py

```diff
@@ -135,25 +157,78 @@ def load_checkpoint(self, state: State):
         """
         seed_to_restore = None
 
-        with tempfile.TemporaryDirectory() as checkpoint_folder:
+        # download the checkpoint to the node-local run directory
+        # since the run directory is sub-foldered by rank,
+        # the node-local run directory is just the parent folder
+        tempdir_ctx = tempfile.TemporaryDirectory() if dist.get_local_rank() == 0 else contextlib.nullcontext(None)
+        with tempdir_ctx as tempdir:
+            local_node_zero = dist.get_local_world_size() * dist.get_node_id()
+            tmpdir_list = [tempdir]
+            dist.broadcast_object_list(tmpdir_list, src=local_node_zero)
```

Comment:
This won't work for multinode - processes will be broadcasting across node boundaries.

The "correct" way to do this would be with [distributed process groups](https://pytorch.org/docs/stable/distributed.html#groups), but this is not supported for NCCL.

Barring that, I think you'll have to use `all_gather_object`, with the non-local-rank-0 nodes passing in `None`.

Reply:
Sounds good, refactored to use `all_gather_object`